### PR TITLE
Update 1.2.7 (For numpy 2.0.0)

### PR DESCRIPTION
**Pull request by JamesRobertsonGames**
mkl_random 1.2.7

**Destination channel:**  defaults

### Links

- [PKG-5572](https://anaconda.atlassian.net/browse/PKG-5572) 
- [Upstream repository](https://github.com/IntelPython/mkl_random)
- [Upstream changelog/diff](https://github.com/IntelPython/mkl_random/releases/tag/v1.2.7)
- Relevant dependency PRs:
  - numpy 2.0 (non-base) for 64bit platforms

### Explanation of changes:

- update tests and update version and numpy-base lower bound and python version 

[PKG-5572]: https://anaconda.atlassian.net/browse/PKG-5572?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -27,7 +29,7 @@ requirements:
     - wheel
     - mkl-devel  {{ mkl }}
     - cython 3
-    - numpy-base {{ numpy }}
+    - numpy-base >=1.16
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]:
```suggestion
    - numpy-base {{ numpy }}
```
mkl_random uses the numpy C API. We have to sue the pinning from aggregate's CBC file.

**Review by JeanChristopheMorinPerso — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
 
 source:
   url: https://github.com/IntelPython/mkl_random/archive/v{{ version }}.tar.gz
-  sha256: 9161eee81fbf27934eac1bc0d26ad3b933a582721fa66fe87e930ff2e9867c52
+  sha256: b6d39f4847921b6603201a491225b086aa75a1c1667ea7068b8813e0514133f7
 
 build:
   number: 0
   skip: true  # [not x86]
+  skip: True  # [py<39 or py>=313]
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]: I don't think the Python 3.7 has been dropped, see https://github.com/IntelPython/mkl_random/blob/v1.2.7/setup.py#L147. But it's true that their CI builds only py39-py312

**Comment by JamesRobertsonGames** [comment]
Just want to make sure about one thing during this review, when changing

numpy-base {{ numpy }}

to 

numpy-base >=1.17

It'll build with 2.0.0 base, which inevitably is what I need for numpy 2.0.0 build on 64 bit platforms, using `numpy-base {{ numpy }}` which is what it once was, it avoids 2.0.0. Should I change anything to ensure that 2.0.0 is being built or is the current implementation okay for what I need

**Review by cbouss — COMMENTED** [review]
This version is to add support for numpy 2.
I would set `skip: True # [py<39]` as there will be no numpy 2 for lower than 3.8 and below and no change in mk_random that justify an update.
Then `numpy-base 2` in host. Then replace pin_compatible in run (which was ineffective due to numpy not being in host) by `numpy >=1.16` to follow upstream. 

Note that the cf recipe is by mkl_random maintainer: https://github.com/conda-forge/mkl_random-feedstock

**Comment by cbouss** [comment]
> Just want to make sure about one thing during this review, when changing
> 
> numpy-base {{ numpy }}
> 
> to
> 
> numpy-base >=1.17
> 
> It'll build with 2.0.0 base, which inevitably is what I need for numpy 2.0.0 build on 64 bit platforms, using `numpy-base {{ numpy }}` which is what it once was, it avoids 2.0.0. Should I change anything to ensure that 2.0.0 is being built or is the current implementation okay for what I need

Yes we need to build with numpy 2

**Comment by JamesRobertsonGames** [comment]
Thanks @cbouss ! Wasn't sure about the logistics for specific version builds when to allow the solver to have its way. Will make those changes now

File: recipe/meta.yaml [hunk 3]
```diff
@@ -27,20 +29,22 @@ requirements:
     - wheel
     - mkl-devel  {{ mkl }}
     - cython 3
-    - numpy-base {{ numpy }}
+    - numpy-base 2.0.0
+    - numpy >=1.16
```
Comment (cbouss) [inline recipe/meta.yaml]:
```suggestion
```
You cannot have numpy here due to the cyclic dependency.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -27,20 +29,22 @@ requirements:
...
   run:
     - python
     - mkl
-    - {{ pin_compatible('numpy') }}
+    - numpy
```
Comment (cbouss) [inline recipe/meta.yaml]:
```suggestion
    - numpy >=1.16
```
To match upstream.

**Comment by JamesRobertsonGames** [comment]
Seems like when the numpy (not base) version is pinned to a lower bound it gets this on osx 

```
conda_libmamba_solver.conda_build_exceptions.ExplainedDependencyNeedsBuildingError: Unsatisfiable dependencies for platform osx-64: {MatchSpec("mkl-devel==2023.1.0=h0e174c6_43559"), MatchSpec("blas=[build=mkl]")}
30 August 2024,04:03:05 	prefect.build	INFO	    Encountered problems while solving:
30 August 2024,04:03:05 	prefect.build	INFO	      - package mkl-devel-2023.1.0-h0e174c6_43559 requires blas * mkl, but none of the providers can be installed
30 August 2024,04:03:05 	prefect.build	INFO	
30 August 2024,04:03:05 	prefect.build	INFO	    Could not solve for environment specs
30 August 2024,04:03:05 	prefect.build	INFO	    The following packages are incompatible
30 August 2024,04:03:05 	prefect.build	INFO	    ├─ mkl-devel 2023.*  is installable and it requires
30 August 2024,04:03:05 	prefect.build	INFO	    │  └─ blas * mkl, which can be installed;
30 August 2024,04:03:05 	prefect.build	INFO	    └─ numpy-base 2.0.0.*  is not installable because it requires
30 August 2024,04:03:05 	prefect.build	INFO	       └─ blas * openblas, which conflicts with any installable versions previously reported.
```

And advice on how to remedy a dependency issue like this?

**Comment by cbouss** [comment]
> Seems like when the numpy (not base) version is pinned to a lower bound it gets this on osx 
> 
> ```
> conda_libmamba_solver.conda_build_exceptions.ExplainedDependencyNeedsBuildingError: Unsatisfiable dependencies for platform osx-64: {MatchSpec("mkl-devel==2023.1.0=h0e174c6_43559"), MatchSpec("blas=[build=mkl]")}
> 30 August 2024,04:03:05 	prefect.build	INFO	    Encountered problems while solving:
> 30 August 2024,04:03:05 	prefect.build	INFO	      - package mkl-devel-2023.1.0-h0e174c6_43559 requires blas * mkl, but none of the providers can be installed
> 30 August 2024,04:03:05 	prefect.build	INFO	
> 30 August 2024,04:03:05 	prefect.build	INFO	    Could not solve for environment specs
> 30 August 2024,04:03:05 	prefect.build	INFO	    The following packages are incompatible
> 30 August 2024,04:03:05 	prefect.build	INFO	    ├─ mkl-devel 2023.*  is installable and it requires
> 30 August 2024,04:03:05 	prefect.build	INFO	    │  └─ blas * mkl, which can be installed;
> 30 August 2024,04:03:05 	prefect.build	INFO	    └─ numpy-base 2.0.0.*  is not installable because it requires
> 30 August 2024,04:03:05 	prefect.build	INFO	       └─ blas * openblas, which conflicts with any installable versions previously reported.
> ```
> 
> And advice on how to remedy a dependency issue like this?

Oh right, you can skip OSX as we stopped building with mkl there.

**Review by skupr-anaconda — APPROVED** [review]

**Review by cbouss — APPROVED** [review]

**Pull request merged**